### DISCUSSION TITLE
CHAPTER5 - QnA

### DISCUSSION THREAD

**Discussion by BvrPark** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#54] [CHAPTER 5 스트림 활용 - Q&A]
### 질문 1.
P.157에서 `distinct` 메서드의 고유 여부는 스트림에서 만든 객체의 hashCode, equals로 결정된다고 합니다.
그래서 `distinct` 메서드를 호출하면 각 요소의 `hashCode()`를 계산하고 중복된 요소를 제거하고, 
같은 요소끼리는 `equals()`메서드를 호출해서 동등성 비교를 수행하여 `equals()`가 true를 반환하면 중복되었다고 생각하여 제거한다고 합니다.

1. 그러면 여기서 `equals()`의 결과가 true인 객체들은 같은 해시코드를 가져야 되는데 만약 다른 해시 코드를 반환할 수도 있을까요? 
-> 예를 들어 해시충돌과 같은 이유 때문에…?

2. 그리고 `distinct`를 사용하는 컬렉션의 크기가 큰 경우에는 그 구성 요소마다 `hashCode()`와 `equals()` 메서드를 하나하나 다 호출해서 사용하기 때문에 좀 시간이 오래 걸리고 비효율적일 것 같은데 혹시 이럴 경우엔 다른 방법이 있나요?

</br>

### 질문 2.
P. 163 ~ 164의
```java
words.stream()
.map(word -> word.split(""))
.distinct()
.collect(toList());
```

1. 위의 코드에서
map(word -> word.split(“”)) 가 `Hello Word`라는 글자를
[“H”], [“e”], [“l”], [“l”], [“o”], [“W”], [“o”], [“r”], [“d”]
이런 식으로 배열에 하나하나 들어가서 `distinct()`가 적용이 안되는 것이라고 이해를 했는데 제가 올바르게 이해를 한건가요?

2. 1번에 제가 이해한 것이 맞다면 `flatMap(Arrays::stream)`은 저렇게 나뉘어진 배열을
“Hello World”라는 String의 형태로 만들어주는 것이 맞나요?

</br>

### 질문 3
기존의 반복을 이용해서 합이나 count를 하는 것이랑 reduce를 이용한 것의 차이는

reduce를 이용하면 내부반복이 추상화되면서 내부 구현에서 병렬로 reduce를 실행할 수 있게 되지만 
기존의 반복을 이용한 합에서는 sum 변수를 공유해야 하므로 쉽게 병렬화하기 어렵다는 차이가 있다라는 것을 알았습니다.

1. 제가 말한 차이 외에도 혹시 다른 점들이 있을까요?
2. 그리고 가변 누적자 패턴이라고 가변 상태를 사용해서 누적 작업을 수행하는 패턴이 있던데 여기서 질문이 있습니다.
그렇다면 굳이 함수형 프로그래밍 대신 일반적인 자바 코드 형태로 사용해서 가변 상태를 사용하면 되는데 왜 가변 누적자 패턴이라는 것을 만들고 사용하는 것인지 궁금해서 다른 분들의 의견을 듣고 싶어서 질문을 올렸습니다.

**Comment by JoisFe** [comment]
A1 - 1) 이 부분은 자바의 java.util.HashMap 혹은 HashSet이 어떻게 구현되어 있는지 보시면 좋을 것 같습니다.
예전에 정리해둔 자료가 있는데 참고하시면 좋을 것 같습니다. [참고 자료](https://github.com/orgs/Study-2-Effective-Java/discussions/23)
A1 -2) 만약 객체를 비교하는 상황이라면 eqauls 정의에 따라 어떤 기준으로 비교를 할 것인지 판단하기 때문에 equals를 사용하지 않고는 비교를 한다는게 쉽지 않을 것 같습니다. (equals만 쓰면되지 왜 Hashcode도 비교하지? -> A1 - 1 의 참고자료를 보시면 이해되실 것 같습니다.)

A 3 - 2) 일단 가변 누적자 패턴에서 가변은
`가변 누적자 패턴(Variadic Accumulator Pattern)은 자바에서 메서드에 가변 개수의 인수를 전달하여 누적하는 패턴을 말합니다. 이 패턴은 메서드가 다양한 개수의 인수를 처리해야 할 때 유용하게 사용`
특정 가변 변수를 사용하여 동시성에 문제를 일으키는 (가변 변수를 사용) 의미는 아닌 것 같습니다. 책에서 사용하는 코드를 보더라도 
``` java
~.parallelStream.reduce(0 -> Integer::sum)
```

여기서 가변 변수는 보이지 않습니다.
- reduce 함수가 내부적으로 동시성에 문제를 일으키지 않도록 불변성을 지키지 않을까 싶습니다.

**Comment by corock** [comment]
## 질문 2

> 1. 위의 코드에서
> map(word -> word.split(“”)) 가 Hello Word라는 글자를
> [“H”], [“e”], [“l”], [“l”], [“o”], [“W”], [“o”], [“r”], [“d”]
> 이런 식으로 배열에 하나하나 들어가서 distinct()가 적용이 안되는 것이라고 이해를 했는데 제가 올바르게 이해를 한건가요?

배열은 참조 자료형이므로 주소가 각각 다르기 때문에 적용이 안 되지 않지 않나 싶습니다.

<br>

> 2. 1번에 제가 이해한 것이 맞다면 flatMap(Arrays::stream)은 저렇게 나뉘어진 배열을
> “Hello World”라는 String의 형태로 만들어주는 것이 맞나요?

`"Hello World"` 라는 문자열 형태라기보다는 위 예시의 경우 각 원소를 평평하게 스트림으로 펼쳐놓았다고 보면 될 것 같습니다.

<br>

> 3. 그리고 간단하게 말하면 map과 flatMap의 차이는
> map은 1:1로 결과를 반환해주는 것이고
> flatMap은 1:N으로 결과를 반환해주는 것이라고 이해를 했는데 이것도 맞나요?

질문을 정확하게 이해 못했어요 ㅠㅠ

**Comment by Irisation23** [comment]
2023-07-04

@crupy 님이 질문 주셨던 3.2 번째 질문에 대한 참조 자료 입니다.
- https://leejaedoo.github.io/working_with_streams/
- by @CoRock
